store/mongo/migration_1_2_15.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package mongo
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/mendersoftware/go-lib-micro/mongo/migrate"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	mopts "go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/mendersoftware/deployments/model"
+)
+
+type migration_1_2_15 struct {
+	client *mongo.Client
+	db     string
+}
+
+// Up intrduces special representation of artifact provides and index them.
+func (m *migration_1_2_15) Up(from migrate.Version) error {
+	ctx := context.Background()
+	c := m.client.Database(m.db).Collection(CollectionImages)
+	cr := m.client.Database(m.db).Collection(CollectionReleases)
+
+	// create index for release name
+	storage := NewDataStoreMongoWithClient(m.client)
+	if err := storage.EnsureIndexes(
+		m.db,
+		CollectionReleases,
+		IndexReleaseName,
+	); err != nil {
+		return err
+	}
+
+	cursor, err := c.Find(ctx, bson.M{})
+	if err != nil {
+		return err
+	}
+	defer cursor.Close(ctx)
+	var a model.Image
+	for cursor.Next(ctx) {
+		err = cursor.Decode(&a)
+		if err != nil {
+			break
+		}
+
+		opt := &mopts.UpdateOptions{}
+		upsert := true
+		opt.Upsert = &upsert
+		update := bson.M{
+			"$set": bson.M{
+				StorageKeyReleaseName:     a.ArtifactMeta.Name,
+				StorageKeyReleaseModified: time.Now(),
+			},
+			"$push": bson.M{StorageKeyReleaseArtifacts: a},
+		}
+
+		res, err := cr.UpdateOne(
+			ctx,
+			bson.M{
+				"$and": []bson.M{
+					{
+						StorageKeyReleaseName: a.ArtifactMeta.Name,
+					},
+					{
+						StorageKeyReleaseArtifactsId: bson.M{
+							"$ne": a.Id,
+						},
+					},
```

Comment:
as I mentioned in the previous comment I was not focusing on optimizing the migration, but on making it possible to run it in the background, but when it comes to artifact id on the list, this field is **not** unique; artifact id in the images collection is an _id, so it is unique, and under normal operations we won't put two images with the same id, but that's not the case for the migration;

Reply:
Aren't the two the same? We're pushing documents from the images collection into a release document, hence the uniqueness constraint from the images collection carries over?

Reply:
if someone will upload new artifact when we'll run the new version of the service, but before (or while) we'll run the migration, then the image will be both in the images and releases (because we're running new version);
during the migration we'll try to copy all documents from images collection to releases, but images uploaded before migration will already be there and there is no index on image id field in the releases collection, so the migration will create a duplicate document